File: src/main/kotlin/dev/slne/surf/essentialsold/commands/minecraft/WeatherCommand.java
package dev.slne.surf.essentialsold.commands.minecraft;

import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.jorel.commandapi.wrappers.NativeProxyCommandSender;
import dev.slne.surf.essentialsold.commands.EssentialsCommand;
import dev.slne.surf.essentialsold.utils.EssentialsUtil;
import dev.slne.surf.essentialsold.utils.color.Colors;
import dev.slne.surf.essentialsold.utils.permission.Permissions;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.World;

import java.util.Objects;
import java.util.function.BiConsumer;

public class WeatherCommand extends EssentialsCommand {
    public WeatherCommand() {
        super("weather", "weather <clear | rain | thunder> [<duration>]", "Change game weather");

        withPermission(Permissions.WEATHER_PERMISSION);

        executesNative((NativeResultingCommandExecutor) (sender, args) -> queryWeather(sender));

        then(weatherTypeArgument("weather")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> setWeather(
                        sender,
                        Objects.requireNonNull(args.getUnchecked("weather")),
                        6000
                ))
                .then(timeArgument("duration")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> setWeather(
                                sender,
                                Objects.requireNonNull(args.getUnchecked("weather")),
                                args.getUnchecked("duration")
                        ))
                )
        );
    }

    private int queryWeather(NativeProxyCommandSender source) {
        val callee = source.getCallee();
        val world = source.getWorld();
        val isClear = !world.isThundering() && !world.hasStorm();

        val clearDuration = world.getClearWeatherDuration();
        val rainDuration = world.getWeatherDuration();
        val thunderDuration = world.getThunderDuration();


        if (isClear) {
            EssentialsUtil.sendSuccess(callee, weatherComponent$adventure("Klar", world, clearDuration));
        } else if (world.isThundering()) {
            EssentialsUtil.sendSuccess(callee, weatherComponent$adventure("Gewitter", world, thunderDuration));
        } else {
            EssentialsUtil.sendSuccess(callee, weatherComponent$adventure("Regen", world, rainDuration));
        }
        return 1;
    }

    private int setWeather(NativeProxyCommandSender source, WeatherType weatherType, Integer durationInTicks) {
        val world = source.getWorld();
        weatherType.setWeather(world, durationInTicks);

        EssentialsUtil.sendSuccess(source.getCallee(), Component.text("Das Wetter in der Welt ", Colors.INFO)
                .append(EssentialsUtil.getDisplayName(world))
                .append(Component.text(" wurde auf ", Colors.INFO))
                .append(Component.text(weatherType.name, Colors.VARIABLE_VALUE))
                .append(Component.text(" für ", Colors.INFO))
                .append(Component.text(EssentialsUtil.ticksToString(durationInTicks), Colors.VARIABLE_VALUE))
                .append(Component.text(" gesetzt.", Colors.INFO)));

        return durationInTicks;
    }

    private Component weatherComponent$adventure(String weather, World world, int durationInTicks) {
        return Component.text("Das Wetter in der Welt ", Colors.INFO)
                .append(EssentialsUtil.getDisplayName(world))
                .append(Component.text(" ist ", Colors.INFO))
                .append(Component.text(weather, Colors.VARIABLE_VALUE))
                .append(Component.text(" für ", Colors.INFO))
                .append(Component.text(EssentialsUtil.ticksToString(durationInTicks), Colors.VARIABLE_VALUE));
    }

    /**
     * Weather types
     */
    @Getter
    @RequiredArgsConstructor
    @FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
    public enum WeatherType {
        CLEAR("Klar", (world, duration) -> {
            world.setThundering(false);
            world.setStorm(false);
            world.setClearWeatherDuration(duration);
        }),
        RAIN("Regen", (world, duration) -> {
            world.setThundering(false);
            world.setStorm(true);
            world.setWeatherDuration(duration);
        }),
        THUNDER("Gewitter", (world, duration) -> {
            world.setStorm(true);
            world.setThundering(true);
            world.setThunderDuration(duration);
        });

        String name; // Name of the weather type
        BiConsumer<World, Integer> setWeather; // Function to set the weather

        /**
         * Set the weather in the given world
         *
         * @param world           The world to set the weather in
         * @param durationInTicks The duration of the weather in ticks
         */
        public void setWeather(World world, int durationInTicks) {
            setWeather.accept(world, durationInTicks);
        }
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/minecraft/WeatherCommand.kt
package dev.slne.surf.essentials.command.minecraft

import com.github.shynixn.mccoroutine.folia.globalRegionDispatcher
import com.github.shynixn.mccoroutine.folia.launch
import dev.jorel.commandapi.kotlindsl.*
import dev.slne.surf.api.core.messages.adventure.sendText
import dev.slne.surf.essentials.command.argument.durationArgument
import dev.slne.surf.essentials.command.argument.weatherTypeArgument
import dev.slne.surf.essentials.plugin
import dev.slne.surf.essentials.util.permission.EssentialsPermissionRegistry
import dev.slne.surf.essentials.util.util.ticks
import dev.slne.surf.essentials.util.util.userContent
import dev.slne.surf.essentials.util.weather.WeatherType
import dev.slne.surf.essentials.util.weather.getWeatherType
import org.bukkit.Bukkit
import java.time.Duration

fun weatherCommand() = commandTree("weather") {
    withPermission(EssentialsPermissionRegistry.WEATHER_COMMAND)

    literalArgument("query") {
        playerExecutor { player, _ ->
            player.sendText {
                appendInfoPrefix()
                info("In der Welt ")
                variableValue(player.world.name)
                info(" ist das Wetter aktuell ")
                variableValue(player.world.getWeatherType().displayName)
                info(".")
            }
        }
    }

    weatherTypeArgument("weather") {
        anyExecutor { sender, arguments ->
            val weather: WeatherType by arguments

            plugin.launch(plugin.globalRegionDispatcher) {
                Bukkit.getWorlds().forEach {
                    weather.setWeather(it, 6000)
                }
            }

            sender.sendText {
                appendSuccessPrefix()
                success("Das Wetter in allen Welten wurde zu ")
                variableValue(weather.displayName)
                success(" geändert.")
            }
        }

        durationArgument("duration") {
            anyExecutor { sender, arguments ->
                val weather: WeatherType by arguments
                val duration: Duration by arguments

                plugin.launch(plugin.globalRegionDispatcher) {
                    Bukkit.getWorlds().forEach {
                        weather.setWeather(it, duration.toMillis().ticks())
                    }
                }

                sender.sendText {
                    appendSuccessPrefix()
                    success("Das Wetter in allen Welten wurde zu ")
                    variableValue(weather.displayName)
                    success(" für ")
                    variableValue(duration.userContent())
                    success(" geändert.")
                }
            }
        }
    }
}